lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+
+	// Update the state change file size.
+	transactionLen := uint64(len(writeBytes))
+	unflushedBytes.StateChangeBytesSize += transactionLen
+
+	stateChangeSyncer.UnflushedBytes[flushId] = unflushedBytes
+}
+
+// FlushTransactionsToFile writes the bytes that have been cached on the StateChangeSyncer to the state change file.
+func (stateChangeSyncer *StateChangeSyncer) FlushTransactionsToFile(event *StateSyncerFlushedEvent) error {
+	flushId := event.FlushId
+	// Get the relevant global flush ID from the state change syncer if the flush ID is nil.
+	if event.IsMempoolFlush && event.FlushId == uuid.Nil {
+		flushId = stateChangeSyncer.MempoolFlushId
+	} else if !event.IsMempoolFlush && event.FlushId == uuid.Nil {
+		flushId = stateChangeSyncer.BlockSyncFlushId
+	}
+
+	var flushFile *os.File
+	var flushFileSize uint64
+	var indexFile *os.File
+	if event.IsMempoolFlush {
+		flushFile = stateChangeSyncer.StateChangeMempoolFile
+		flushFileSize = stateChangeSyncer.StateChangeMempoolFileSize
+		indexFile = stateChangeSyncer.StateChangeMempoolIndexFile
+	} else {
+		flushFile = stateChangeSyncer.StateChangeFile
+		flushFileSize = stateChangeSyncer.StateChangeFileSize
+		indexFile = stateChangeSyncer.StateChangeIndexFile
+	}
+
+	// If the flush failed, delete the unflushed bytes and associated metadata.
+	// Also delete any unconnected mempool txns from our cache.
+	if !event.Succeeded {
+		delete(stateChangeSyncer.UnflushedBytes, flushId)
+	}
+
+	unflushedBytes, exists := stateChangeSyncer.UnflushedBytes[flushId]
+
+	if !exists {
+		return nil
+	}
+
+	if unflushedBytes.StateChangeBytes == nil || unflushedBytes.StateChangeOperationIndexes == nil || len(unflushedBytes.StateChangeOperationIndexes) == 0 {
+		return fmt.Errorf("Error flushing state change file: FlushId %v has nil bytes\n", flushId)
+	}
+
+	// Write the encoded StateChangeEntry bytes to the state changer file.
+	_, err := flushFile.Write(unflushedBytes.StateChangeBytes)
+
+	if err != nil {
+		return fmt.Errorf("Error writing to state change file: %v", err)
+	}
+
+	// Buffer to hold bytes for index file
+	stateChangeIndexBuf := make([]byte, 0)
+
+	// Loop through the index bytes of the state change file and write them to the index file.
+	// The StateChangeOperationIndexes array contains the byte index of		 where each transaction occurs within the
+	// unflushed state change bytes (e.g. the first value in the slice will always be 0).
+	// We need to add the size of the state change file to each of these values to get the byte index
+	// in the state change file.
+	for _, indexBytes := range unflushedBytes.StateChangeOperationIndexes {
+		// Get the byte index of where this transaction occurs in the state change file.
+		dbOperationIndex := indexBytes + flushFileSize
+		// Convert the byte index value to a uint64 byte slice.
+		dbOperationIndexBytes := make([]byte, 8)
+		binary.LittleEndian.PutUint64(dbOperationIndexBytes, dbOperationIndex)
```

Comment:
Is there a reason we encode a full uint64 rather than using UintToBuf? I don't think the efficiency matters much just figured I'd mention it.

Reply:
For context, UintToBuf uses the minimum number of bytes to encode the value. So it would be 2-3 ish bytes for each number instead of 8.

Reply:
We need to be able to pinpoint where in the file the pointer lives for a particular index. If we used a dynamic-sized encoding, we'd need to iterate through every pointer until we reach the index we want. By using a standard-sized uint64, we seek straight to index * 8 to find the pointer we want.